### PR TITLE
[clang][AArch64] Add validation for Global Register Variable.

### LINKED ISSUE #76426: [AArch64] Crash when compiling global register variable __asm__("x15") without -ffixed-x15
clang in AArch64 environment, fatal error at compile time.
If a global register variable associated with the register x15 is defined in the program and referenced in the program, a fatal error occurs at compile time when -ffixed-x15 is not specified.
When -ffixed-x15 is specified, it can be compiled.
Though failing to compile without -ffixed-x15 may be intentional, it should not be a fatal error.
Here is the program and compilation log:

[program]
```c
register int i1 __asm__("x15");

int foo(int i) {
  return i1;
}
int main() {
  return 0;
}
```

[compile log]
```
$ clang test_x15.c
fatal error: error in backend: Invalid register name "x15".
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 18.0.0 (https://github.com/llvm/llvm-project.git fe846bc9bd164ed89cef1caefb681e0aee3d4dec)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /work/home/marakawa/bin
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/test_x15-659ab1.c
clang: note: diagnostic msg: /tmp/test_x15-659ab1.sh
clang: note: diagnostic msg:

********************
$
```

[test_x15-659ab1.c]
```
# 1 "<built-in>"
# 1 "test_x15.c"
register int i1 __asm__("x15");

int foo(int i) {
  return i1;
}
int main() {
  return 0;
}
```

[test_x15-659ab1.sh]
```
# Crash reproducer for clang version 18.0.0 (https://github.com/llvm/llvm-project.git fe846bc9bd164ed89cef1caefb681e0aee3d4dec)
# Driver args: "test_x15.c"
# Original command:  "/work/home/marakawa/LLVM/llvm-project/installs/release/bin/clang-18" "-cc1" "-triple" "aarch64-unknown-linux-gnu" "-emit-obj" "-mrelax-all" "-dumpdir" "a-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test_x15.c" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=non-leaf" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "generic" "-target-feature" "+neon" "-target-feature" "+v8a" "-target-abi" "aapcs" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/work/home/marakawa/test/local_test/1179_fatal" "-fcoverage-compilation-dir=/work/home/marakawa/test/local_test/1179_fatal" "-resource-dir" "/work/home/marakawa/LLVM/llvm-project/installs/release/lib/clang/18" "-internal-isystem" "/work/home/marakawa/LLVM/llvm-project/installs/release/lib/clang/18/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/opt/rh/gcc-toolset-12/root/usr/lib/gcc/aarch64-redhat-linux/12/../../../../aarch64-redhat-linux/include" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-ferror-limit" "19" "-fno-signed-char" "-fgnuc-version=4.2.1" "-fcolor-diagnostics" "-target-feature" "+outline-atomics" "-target-feature" "-fmv" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "/tmp/test_x15-518c67.o" "-x" "c" "test_x15.c"
 "/work/home/marakawa/LLVM/llvm-project/installs/release/bin/clang-18" "-cc1" "-triple" "aarch64-unknown-linux-gnu" "-emit-obj" "-mrelax-all" "-dumpdir" "a-" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test_x15.c" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-mframe-pointer=non-leaf" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "generic" "-target-feature" "+neon" "-target-feature" "+v8a" "-target-abi" "aapcs" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/work/home/marakawa/test/local_test/1179_fatal" "-fcoverage-compilation-dir=/work/home/marakawa/test/local_test/1179_fatal" "-ferror-limit" "19" "-fno-signed-char" "-fgnuc-version=4.2.1" "-fcolor-diagnostics" "-target-feature" "+outline-atomics" "-target-feature" "-fmv" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c" "test_x15-659ab1.c"
```

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
Fixes: #76426

### CODE REVIEW

**Comment by kawashima-fj** [comment]
@DanielKristofKiss Thanks for the fix. At first glance, LGTM. I'll take a closer look in a few days.
I'm not so familiar with the code. If someone else can review, I'll leave it to them.

File: clang/test/Driver/aarch64-fixed-register-global.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Check that -ffixed register handled for globals.
+// Regression test for #76426
+// RUN: %clang --target=aarch64-none-gnu -ffixed-x15 -### %s 2>&1 | FileCheck %s
+// CHECK-NOT: fatal error: error in backend: Invalid register name "x15".
```
Comment (kawashima-fj) [inline clang/test/Driver/aarch64-fixed-register-global.c:4]: You modified Clang code but this error message is the one of the LLVM backend.
Is it better to check the Clang-side message `error: size of register 'x15' does not match variable size`?

**Review by kawashima-fj — APPROVED** [review]
Thanks for the fix. I left a minor suggestion.

Some notes:

- This commit make Clang emit `error: size of register '???' does not match variable size` error instead of `fatal error: error in backend: Invalid register name "???"`, regardless of the `-ffixed-???` option.
- #76426 says "When `-ffixed-x15` is specified, it can be compiled.". When `-ffixed-x15` is specified, older revisions (including 18.1.x releases) compiles the program in #76426 without an error but the generated code is incorrect (the `foo` function is compiled into `mrs w0, NZCV; ret`!). Recent revisions emits `fatal error: error in backend: Invalid register name "x15".`. So make it a Clang error is appropriate.
- GCC allows this variable-register size mismatch. Clang does not allow.
- We don't need to use `starts_with_insensitive`. Register names are case-sensitive (same as GCC).

File: clang/test/Driver/aarch64-fixed-register-global.c (hunk 1 above)
Comment (DanielKristofKiss) [reply clang/test/Driver/aarch64-fixed-register-global.c:4]: Right! I'd keep the bugreport as is for regression test but added a new tests for the reg check.

**Pull request merged**